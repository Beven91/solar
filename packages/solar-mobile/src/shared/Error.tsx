import './index.scss';
import React from 'react';
import Crash from '../crash';

export interface PageErrorProps {
  // 异常标题
  title?: string,
  // 异常消息
  message?: string,
  // 自定义按钮点击
  onClick?: () => void,
}

export default class PageError extends React.Component<PageErrorProps, {}> {
  static defaultProps = {
    onClick: () => history.back(),
  }

  render() {
    return (
      <Crash { ...this.props } />
    );
  }
}
