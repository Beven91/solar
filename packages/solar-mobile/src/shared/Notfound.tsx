import './index.scss';
import React from 'react';
import Crash from '../crash';

export interface NotfoundProps {
  // 异常标题
  title?: string,
  // 异常消息
  message?: string,
  // 自定义按钮点击
  onClick?: () => void,
}

export default class Notfound extends React.Component<NotfoundProps, {}> {
  static defaultProps = {
    title: '这是一个未知的领域...',
  }

  render() {
    return (
      <Crash {...this.props} />
    );
  }
}
