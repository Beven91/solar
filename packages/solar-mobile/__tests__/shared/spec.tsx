import React from 'react';
import Error from '../../src/shared/Error';
import Notfound from '../../src/shared/Notfound';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';


describe('Crash', () => {
  afterAll(() => {
    location.hash = '';
  });

  it('renders not-found', () => {
    const tree = mount(
      <Notfound />
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('renders not-found title', () => {
    const tree = mount(
      <Notfound title="404页面" />
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('renders not-found back', () => {
    const tree = mount(
      <Notfound />
    );

    const custBack = jest.fn();
    const spyBack = jest.spyOn(history, 'back');
    // 模拟点击
    tree.find('.crash-button').first().simulate('click');
    // 断言：点击后退时，默认调用history.back
    expect(spyBack).toHaveBeenCalledTimes(1);

    // 自定义返回
    tree.setProps({ onClick: custBack });
    // 清除调用记录
    spyBack.mockReset();
    // 模拟点击
    tree.find('.crash-button').first().simulate('click');
    // 断言：当传入自定义后退事件时，仅会触发自定义的onBack
    expect(custBack).toHaveBeenCalledTimes(1);
    // history.back 不会被调用，所以调用次数为0
    expect(spyBack).toHaveBeenCalledTimes(0);
  });


  it('renders error', () => {
    const tree = mount(
      <Error />
    );
    expect(toJson(tree)).toMatchSnapshot();
  });


  it('renders error title', () => {
    location.hash = '';
    const tree = mount(
      <Error title="出错啦..." />
    );
    expect(toJson(tree)).toMatchSnapshot();
  });


  it('renders error back', () => {
    const tree = mount(
      <Error />
    );

    const custBack = jest.fn();
    const spyBack = jest.spyOn(history, 'back');
    // 模拟点击
    tree.find('.crash-button').first().simulate('click');
    // 断言：点击后退时，默认调用history.back
    expect(spyBack).toHaveBeenCalledTimes(1);

    // 自定义返回
    tree.setProps({ onClick: custBack });
    // 清除调用记录
    spyBack.mockReset();
    // 模拟点击
    tree.find('.crash-button').first().simulate('click');
    // 断言：当传入自定义后退事件时，仅会触发自定义的onBack
    expect(custBack).toHaveBeenCalledTimes(1);
    // history.back 不会被调用，所以调用次数为0
    expect(spyBack).toHaveBeenCalledTimes(0);
  });
});
