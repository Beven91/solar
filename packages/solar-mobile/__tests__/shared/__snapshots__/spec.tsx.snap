// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crash renders error 1`] = `
<PageError
  onClick={[Function]}
>
  <Crash
    btnText="返回"
    className=""
    icon=""
    message="大侠别急，您可点击以返回试试~"
    onClick={[Function]}
    title="页面竟然崩溃了"
  >
    <div
      className="crash-component  "
    >
      <img
        className="logo"
        src=""
      />
      <div
        className="crash-title"
      >
         
        页面竟然崩溃了
      </div>
      <div
        className="crash-description"
      >
        大侠别急，您可点击以返回试试~
      </div>
      <div
        className="crash-button"
        onClick={[Function]}
      >
         
        返回
         
      </div>
    </div>
  </Crash>
</PageError>
`;

exports[`Crash renders error title 1`] = `
<PageError
  onClick={[Function]}
  title="出错啦..."
>
  <Crash
    btnText="返回"
    className=""
    icon=""
    message="大侠别急，您可点击以返回试试~"
    onClick={[Function]}
    title="出错啦..."
  >
    <div
      className="crash-component  "
    >
      <img
        className="logo"
        src=""
      />
      <div
        className="crash-title"
      >
         
        出错啦...
      </div>
      <div
        className="crash-description"
      >
        大侠别急，您可点击以返回试试~
      </div>
      <div
        className="crash-button"
        onClick={[Function]}
      >
         
        返回
         
      </div>
    </div>
  </Crash>
</PageError>
`;

exports[`Crash renders not-found 1`] = `
<Notfound
  title="这是一个未知的领域..."
>
  <Crash
    btnText="返回"
    className=""
    icon=""
    message="大侠别急，您可点击以返回试试~"
    onClick={[Function]}
    title="这是一个未知的领域..."
  >
    <div
      className="crash-component  "
    >
      <img
        className="logo"
        src=""
      />
      <div
        className="crash-title"
      >
         
        这是一个未知的领域...
      </div>
      <div
        className="crash-description"
      >
        大侠别急，您可点击以返回试试~
      </div>
      <div
        className="crash-button"
        onClick={[Function]}
      >
         
        返回
         
      </div>
    </div>
  </Crash>
</Notfound>
`;

exports[`Crash renders not-found title 1`] = `
<Notfound
  title="404页面"
>
  <Crash
    btnText="返回"
    className=""
    icon=""
    message="大侠别急，您可点击以返回试试~"
    onClick={[Function]}
    title="404页面"
  >
    <div
      className="crash-component  "
    >
      <img
        className="logo"
        src=""
      />
      <div
        className="crash-title"
      >
         
        404页面
      </div>
      <div
        className="crash-description"
      >
        大侠别急，您可点击以返回试试~
      </div>
      <div
        className="crash-button"
        onClick={[Function]}
      >
         
        返回
         
      </div>
    </div>
  </Crash>
</Notfound>
`;
